File: solutions/c/armstrong-numbers/1/armstrong_numbers.c
#include "armstrong_numbers.h"

bool is_armstrong_number(int candidate) {

    int length = 0;
    int number = candidate;
  
    do {
        length++;
        number =  number / 10;
    } while ( number != 0 );
  
    int sum = 0;
    number = candidate;
  
    for ( int i = 1; i <= length; i++ ) {
        int last_digit = number % 10;
        int pow = last_digit;
        for ( int j = 1; j < length; j++ ) {
            pow *= last_digit;
        }
        sum += pow;
        number = number / 10;
    }
  
    bool test = false;
    if ( sum == candidate ) {
        test = true;
    }
  
    return test;
}

File: solutions/lua/armstrong-numbers/1/armstrong-numbers.lua
local ArmstrongNumbers = {}

function ArmstrongNumbers.is_armstrong_number(number)
    local digits = math.floor(math.log(math.abs(number),10) + 1)
    if number == 0 then
        digits = 1
    end

    local n    = number
    local calc = 0
    while n > 0 do
        local r = n % 10
        calc = calc + r ^ digits
        n = n // 10
    end

    if calc == number then
        return true
    end
    return false
end

return ArmstrongNumbers
